shapes/main.cpp

```diff
@@ -1,54 +1,62 @@
+#include <algorithm>
+#include <functional>
 #include <iostream>
+#include <map>
+#include <memory>
 #include <vector>
-#include <algorithm>
 #include <string>
-#include <memory>
-#include "Shape.hpp"
+#include <type_traits>
+
+#include "Circle.hpp"
 #include "Rectangle.hpp"
+#include "Shape.hpp"
 #include "Square.hpp"
-#include "Circle.hpp"
 
 using namespace std;
 
 using Collection = vector<shared_ptr<Shape>>;
+using mapCollection = map<shared_ptr<Shape>, double>;
+using shapePtr = shared_ptr<Shape>;
 
-bool sortByArea(shared_ptr<Shape> first, shared_ptr<Shape> second)
-{
-    if(first == nullptr || second == nullptr)
-        return false;
+
+auto sortByArea = [](shapePtr first, shapePtr second) {
+    if(first == nullptr || second == nullptr) { return false; }
     return (first->getArea() < second->getArea());
-}
+};
 
-bool perimeterBiggerThan20(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getPerimeter() > 20);
+auto perimeterBiggerThan20 = [](shapePtr s) {
+    if (s) { return (s->getPerimeter() > 20); }
     return false;
-}
+};
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
+std::function<bool(shapePtr)> areaLessThanX = [x = 10](shapePtr s) {
+    if (s) { return s->getArea() < 10; }
     return false;
+};
+
+void printCollectionElements(const Collection& collection)
+{
+    for (const auto& it: collection)
+        if(it)
+            (it)->print();
 }
 
-void printCollection(const Collection& collection)
+void printMapCollectionElements(const mapCollection& collection)
 {
-    for (const auto & it : collection)
-        if (it)
-            it->print();
+    for (const auto& [key, value]: collection)
+        if(key)
+            (key)->print();
 }
 
 void printAreas(const Collection& collection)
 {
-    for (const auto & it : collection)
-        if (it)
-            cout << it->getArea() << std::endl;
+    for (const auto& it: collection)
+        if(it)
+            cout << (it)->getArea() << std::endl;
```

Comment:
```suggestion
    for (const auto& it: collection) {
        if(it) {
            cout << (it)->getArea() << std::endl;
        }
    }
```

Reply:
If we already use std, let's be consistent. Let's also write std::cout

Reply:
done